homeassistant/components/stream/worker.py

```diff
@@ -160,37 +160,34 @@ def check_flush_part(self, packet: av.Packet) -> None:
             for stream_output in self._outputs_callback().values():
                 stream_output.put(self._segment)
         else:  # These are the ends of the part segments
-            self._segment.parts.append(
-                Part(
-                    duration=float(
-                        (packet.dts - self._part_start_dts) * packet.time_base
-                    ),
-                    has_keyframe=self._part_has_keyframe,
-                    data=self._memory_file.getbuffer()[
-                        self._segment_last_write_pos : byte_position
-                    ].tobytes(),
-                )
-            )
+            self.flush_part(packet, byte_position)
+            # We could put the next two lines in self.flush_part, but they are
+            # already done by reset() in the case of flushing the stub part,
```

Comment:
I was playing around with this code more to make sure i understood how it worked in https://github.com/allenporter/home-assistant-core/commit/c4ca6f91fbf732c55d74391bd50980c8b457c658 and had this idea (that I think is overkill) of adding some `BytesIOCursor` class to simplify the `SegmentBuffer` logic.  I don't think my diff is a good idea, but I think this comment here does touch on something interesting.

How about this:
1. Rename `_segment_last_write_pos` to `_memory_file_pos`
2. Everywhere that `_memory_file` is read also advance `_memory_file_pos` (so, move it into `flush_part` and up next to where `Segment` is created
3. Duplicate the `_part_start_dts` update logic next to where `Segment` is created, e.g. `self._part_start_dts = self._segment_start_pos` then also in `flush_part`.

That is, `_part_start_dts` belongs next to wherever the part is created.


Alternatively, maybe we want to go back to having `Segment` created in `reset` like you used to have it before i asked you to change it, and still just update `init` above rather than creating the `Segment`.

Reply:
Let me try that

Reply:
>     2. Everywhere that `_memory_file` is read also advance `_memory_file_pos` (so, move it into `flush_part` and up next to where `Segment` is created
> 
Isn't it already advanced where `Segment` is created?
>     3. Duplicate the `_part_start_dts` update logic next to where `Segment` is created, e.g. `self._part_start_dts = self._segment_start_pos` then also in `flush_part`.
> 
> 
> That is, `_part_start_dts` belongs next to wherever the part is created.
> 

Are you basically saying to just put those two in `flush_part` not really worrying about duplicating the logic for the stub?

> Alternatively, maybe we want to go back to having `Segment` created in `reset` like you used to have it before i asked you to change it, and still just update `init` above rather than creating the `Segment`.

I guess both ways have their advantages. Creating it in reset keeps the reset logic a little cleaner, while creating it after init helps by not having to worry about a `Segment` without an `init`.

Reply:
I did the renaming, but I'm not sure exactly what you mean in 2 and 3...it looks to me like that is just choosing the alternative option suggested by the comment, unless I'm missing something

Reply:
Doesn't seem like there's a clean way to take care of the duplicated logic, but practically it doesn't matter at all.
I moved those lines into flush_part with a comment.
I also changed the Part construction to just use BytesIO.read() - since we are now only reading with it in one place, it's not too messy to use.

Reply:
I ended up making a change to move all the `Part` stuff out of reset (I think you were suggesting this) and adding a `bool` parameter to `flush_part` to avoid duplicated logic. Let me know if you think it's too messy.